### ISSUE TITLE
[no-type-alias] Unhandled in union types are not allowed for `keyof` or `typeof`

### ISSUE THREAD

**Issue by mscharley** [typescript-eslint/typescript-eslint#3202]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
        "@typescript-eslint/no-type-alias": [
          "error",
          {
            "allowAliases": "in-unions-and-intersections",
            "allowLiterals": "in-unions-and-intersections",
            "allowCallbacks": "always",
            "allowConditionalTypes": "always",
            "allowConstructors": "always",
            "allowMappedTypes": "always",
            "allowTupleTypes": "always"
          }
        ],
  }
}
```

```TS
// your repro code case
const SCALARS = {
  Jwt: {}, // Some value here, I don't think it matters for the below code.
} as const;
const FixedNames = {
  Jwt: 'JWT',
} as const;
type FixedNames = typeof FixedNames extends Record<string, infer U> ? U : never;
type ScalarGenerator = (container: interfaces.Container) => GraphQLScalarType;
// Error is triggered on this line.
type ScalarNames = FixedNames | keyof typeof SCALARS;
```

```jsonc
// tsconfig.json
{
  "compilerOptions": {
    "target": "ES2017",
    "lib": [
      "es5",
      "es2015",
      "es2016",
      "es2017",
      "dom",
      "esnext.asynciterable"
    ],
    "module": "commonjs",
    "moduleResolution": "Node",
    "typeRoots": ["./types", "./node_modules/@types"],
    "strict": true,
    "emitDecoratorMetadata": true,
    "experimentalDecorators": true,
    "noUnusedLocals": true,
    "noUnusedParameters": true,
    "resolveJsonModule": true,
    "esModuleInterop": true,
    "sourceMap": true,
    "rootDir": "./src",
    "outDir": "./dist"
  },
  "include": [
    "./src/**/*",
    "./types/**/*.d.ts"
  ]
}
```

**Expected Result**

I would expect this type of construct to either be allowed or be configurable.

**Actual Result**

"Unhandled in union types are not allowed"

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.18.0` |
| `@typescript-eslint/parser`        | `4.18.0` |
| `TypeScript`                       | `4.2.3` |
| `ESLint`                           | `7.22.0` |
| `node`                             | `12.13.1` |
